File: pwh.cpp
// 2.1 Million Guesses Per Second
// aid! 0.260845 in 5.7 Million
#include <iostream>
#include <string>
#include <vector>
#include <chrono>
#include <openssl/sha.h>
#include <sstream>
#include <iomanip>

using namespace std;
using namespace chrono;

// Function to generate SHA256 hash of a string
string sha256(const string& str) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, str.c_str(), str.length());
    SHA256_Final(hash, &sha256);
    stringstream ss;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        ss << hex << setw(2) << setfill('0') << static_cast<int>(hash[i]);
    }
    return ss.str();
}

// Function to generate combinations using recursive backtracking
void generateCombinations(const string& charSet, int maxLength, const string& password, string currentString, int& guesses) {
    static auto start = steady_clock::now();

    if (sha256(currentString) == password) {
        auto end = steady_clock::now();
        double elapsed_seconds = duration<double>(end - start).count();
        cout << "Found " << password << "(" << currentString << ")" << " with " << guesses << " guesses" << " in " << elapsed_seconds << " seconds" << endl;
        exit(0);
    }

    cout << sha256(currentString) << "\n";

    for (char c : charSet) {
        currentString.push_back(c);
        ++guesses;

        if (currentString.length() <= maxLength) {
            generateCombinations(charSet, maxLength, password, currentString, guesses);
        }

        currentString.pop_back();
    }
}

int main() {
    string charSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+[]{}|,'.<>?~";

    string password;
    int maxLength;

    cout << "Hashed Input Password: ";
    cin >> password;

    cout << "Length Before Hashing: ";
    cin >> maxLength;

    string currentString;
    int guesses = 0;

    generateCombinations(charSet, maxLength, password, currentString, guesses);

    return 0;
}

File: python/pwh.py
import hashlib
import time

# Function to generate SHA256 hash of a string
def sha256(s):
    return hashlib.sha256(s.encode()).hexdigest()

# Function to generate combinations using recursive backtracking
def generate_combinations(char_set, max_length, password, current_string, guesses):
    if sha256(current_string) == password:
        elapsed_time = time.time() - generate_combinations.start_time
        print(f"\n\n\n\nFound {password} ({current_string}) with {guesses} guesses in {elapsed_time:.6f} seconds")
        exit(0)

    for c in char_set:
        current_string += c
        guesses += 1

        if len(current_string) <= max_length:
            generate_combinations(char_set, max_length, password, current_string, guesses)

        current_string = current_string[:-1]

def main():
    char_set = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+[]{}|,'.<>?~"

    password = input("Hashed Input Password: ")
    max_length = int(input("Length Before Hashing: "))

    current_string = ""
    guesses = 0

    generate_combinations.start_time = time.time()
    generate_combinations(char_set, max_length, password, current_string, guesses)

if __name__ == "__main__":
    main()
